### ISSUE TITLE
Couchbase lite iOS memory consumption indexing views

### ISSUE THREAD

**Issue by opicciotto** [couchbase/couchbase-lite-ios#701]
We have  reproducible case for crashing CBL iOS with a 13K document database.
This app is a standard PhoneGap app using the CBL plugin 1.04; We made it the simplest possible to be sure no other component was interfering with memory consumption.
You can download it from here :

http://d2lm6861ur7e1f.cloudfront.net/webrepository/DebugMemory.zip

This a a cordova /phone app you can use to build with Phonegap Build or locally. The app will connect to a cloud based CouchDB server .

You can use the "Sync" button to trigger the sync downloading all the database to the mobile. Then you can use the "View" button to launch a query on the database using the javascript view. The view will start indexing and the app crashes about 10 to 15 secs later.

This has been tested on an iPad mini iOS8.1 and an iPhone 4 iOS 8.1

**Comment by pasin** [comment]
Thanks for reporting this. I will look at the issue shortly.

**Comment by pasin** [comment]
Reference: https://forums.couchbase.com/t/couchbase-lite-ios-memory-consumption/3578

Some info taken from the forum.

#Document: 13K

Map function:
```Javascript
function (doc) {
// Get Products AND Categories by levelIdFather
if (doc.shopCode && doc.levelIdFather) {
	if (doc.levelId) { // Category
		emit([doc.shopCode, doc.levelIdFather], {name : doc.name, levelId : doc.levelId});
	} else if (doc.sku) { // Product
		if (typeof doc.levelIdFather == "object") {
			for (var key in doc.levelIdFather) {
		        emit([doc.shopCode, key], {sku : doc.sku, name : doc.name, packaging : doc.packaging, priceOfUnit : doc.priceOfUnit, unitPrice : doc.unitPrice, base_url : doc.base_url, imageNumber : doc.imageNumber});
			}
		} else {
			emit([doc.shopCode, doc.levelIdFather], {sku : doc.sku, name : doc.name, packaging : doc.packaging, priceOfUnit : doc.priceOfUnit, unitPrice : doc.unitPrice, base_url : doc.base_url, imageNumber : doc.imageNumber});
		}
	}
}
```

**Comment by pasin** [comment]
I tested the app on an iPhone 6 with iOS 8.3 and the app crashed with memory warning while waiting for the query (or indexing) to return the result. Look like the lot of memory is hold by JS Garbage Collector.


![instrument](https://cloud.githubusercontent.com/assets/801454/7715517/e406c66e-fe3c-11e4-904f-1c5b995c8e8c.png)

**Comment by pasin** [comment]
I created a simple native iOS app to isolate the issue from being a PhoneGap app, and I was able to get the same issue. The issue seems to be related to when JSC is parsing the data passed from Objective-C to Javascript call but I still have no clues what is causing that.

![stacetrack](https://cloud.githubusercontent.com/assets/801454/7734335/23556c8c-feea-11e4-9585-5c8dda06283b.png)

Note: When I run the same tests with my own ata set which has similar data size but simpler structure (no nested dicts or arrays), I didn't see the issue.

**Comment by snej** [comment]
I can think of three possibilities —

1. Our JS glue code is leaking JS object references
2. View indexing is causing JS objects to get created faster than the GC can clean them up
3. Something about invoking JS from native code requires us to invoke GC manually

2 and 3 seem quite unlikely. Sounds like we should code-review CBLJSFunction and CBLJSViewCompiler...

**Comment by snej** [comment]
I've looked through our code and haven't found any leaks. The only JavaScriptCore types that need explicit releases are JSStringRef, JSClassRef, JSGlobalContextRef, JSContextGroupRef. Of those we only use JSStringRef and JSGlobalContextRef, and we release all the ones we allocate.

Just FYI, there is an Obj-C based JS API available since iOS 7, and it looks like it's not only easier to use, but should make it more efficient for us to convert between Obj-C and JS objects (currently we resort to using JSON strings as an intermediary.) That should speed up view indexing.

**Comment by pasin** [comment]
The new Objective-C API looks a lot simpler. I will play with the new API.

Anyway, I think we have cleaned up all of the objects well in general. When I ran profiling on the simulator (no memory limit so no crash), I was still seeing JS GC Collector object piling up, but after the index was done, all of the JS objects got cleaned up.

The weird thing is that when I indexed another set of data (10K docs with similar doc size), there was no JS garbage collector piling up at all. While using the user's data, only about 3700 docs could be indexed.

**Comment by snej** [comment]
The issue might only happen when calling emit with complex values?

**Comment by pasin** [comment]
The issue happens even with `function(doc) {emit(doc._id, null);}` so emitting the value does not cause the issue here. 

Based on the profiling result, I believe the issue is at https://github.com/couchbase/couchbase-lite-ios/blob/b3b64f38b6278f39b5e90362906f443100e9f536/Source/CBLJSFunction.m#L155 when it converts a dictionary to JSValue object. Maybe it couldn't do GC fast enough for some JSON data.

Anyway, good news is that after I did a quick refactoring to use the new Objective-C APIs, the problem is gone. I don't see the memory issue anymore, and I *feel* the indexing is much faster as well. Going to do a proper refactoring.

**Comment by pasin** [comment]
@opicciotto I have fixe the issue, and the fixed has been merged to the release/1.1.0 branch. While we are waiting for the next release build, you could try to build a new binary yourself and update your PhoneGap plugin. You can follow the build instruction here (https://github.com/couchbase/couchbase-lite-ios/wiki/Building-Couchbase-Lite). Anyway thanks a lot for reporting the issue and the test app /data.

**Comment by opicciotto** [comment]
@pasin, we will do that next week, thanks for fixing
